### ISSUE TITLE
frame count discrepancy

### ISSUE THREAD

**Issue by TobiasPfeifer** [X-LANCE/LSCodec-Inference#2]
### Description

While testing LSCodec, I ran into a frame count discrepancy between the expected and actual output.

**Input:**  
156800 audio samples (9.80 s at 16kHz)

**Expected frame count**
LSCodec: **245 frames**

**Observed output:**
[LSCodec] means shape: torch.Size([1, 243, 64])
[LSCodec] logvars shape: torch.Size([1, 243, 64])
[LSCodec] embed_index shape: torch.Size([243, 1])

It seems that 2 frames are missing compared to the expected count.
For a frame rate of 25 Hz on 16 kHz audio, each frame spans $$1/25$$ seconds, which equals 0.04 s, so the samples per frame are $$16000 \times 0.04 = 640$$ samples per frame.

Given an input of 156800 samples at 16 kHz, the total duration is $$156800 / 16000 = 9.8$$ s, and the expected number of 25 Hz frames is $$9.8 \times 25 = 245$$ frames, equivalently $$156800 / 640 = 245$$ frames.

LSCodec explicitly supports a 25 Hz token/frame rate configuration, so the above conversion is consistent with its stated 25 Hz mode; therefore an output of 243 frames indicates a shortfall of 2 frames relative to the ideal non-overlapping framing of this clip length.

### Steps to Reproduce
- Input an audio clip with 156800 samples (~9.8 seconds)
- Run it through the LSCodec encoder
- Compare the resulting frame count with the expected number

### Expected Behavior
Frame count should match the expected 245 frames.

### Actual Behavior
LSCodec outputs only 243 frames

**Comment by cantabile-kwok** [comment]
Hi, sorry for the late reply (somehow I did not receive a notification for new issues).
This frame count issue is a known issue that stems from the conv1d operations in the encoder. Although it is expected that the model produces exactly 50Hz or 25Hz tokens, in practice the frame rate is determined by a stack of 1D convolutions. As each convolution layer introduces an integer flooring operation, the stack of them could bring this mismatch. Here is a detailed analysis of this behavior:

From the official torch documentation of conv1d (https://docs.pytorch.org/docs/stable/generated/torch.nn.Conv1d.html), we can easily calculate the output length given the input length. Here,all the conv layers use dilation=1 and padding=0, so the equation becomes only relevant to the strides and kernel sizes. Through simple derivations we could find `output_length=int(input_length/stride)-1` for each of the conv1d operation in LSCodec encoder. Let's take the 50hz version for example, there are 5 conv1d with strides (5,4,4,2,2). Then the total output length can be determined by the following image. So, although the output should be 50 tokens for an input of 1s in 16000Hz, the actual output is 48 tokens.

<img width="526" height="404" alt="Image" src="https://github.com/user-attachments/assets/48a86d90-6913-4835-b559-657b8edc2c5c" />

We call it 50Hz and 25Hz as an approximate, because given the conv1d properties, the output frame rate is not a constant in practice. This is also very common in other codec models in my experience. If you require strictly 25Hz outputs, feel free to pad on your input or output; otherwise, it is a known and expected "feature".

**Comment by TobiasPfeifer** [comment]
thanks for the detailed explanation
